finetuner/tuner/paddle/miner.py

```diff
@@ -97,6 +194,82 @@ def mine(
             return _empty_tensor(), _empty_tensor(), _empty_tensor()
 
 
+class TripletEasyHardMiner(BaseClassMiner[paddle.Tensor]):
+    def __init__(self, pos_strategy: str = 'hard', neg_strategy: str = 'hard'):
+        """
+        Miner implements easy-hard mining for triplets during training with
+        triplet loss. The following strategies are available.
+
+        Pos. Strategy:
+        - 'hard': Returns hardest positive (furthest) sample per anchor
+        - 'semihard': Returns the hardest positive sample per anchor, such
+          that it is closer than the selected negative
+        - 'easy': Returns the easiest positive sample per anchor
+        - 'all': Returns all positive samples
+
+        Neg. Strategy:
+        - 'hard': Returns hardest negative (closest) sample per anchor
+        - 'semihard': Returns the hardest negative sample per anchor, such
+          that it is further than the selected negative
+        - 'easy': Returns the easiest negative sample per anchor
+        - 'all': Returns all negative samples
+
+        Not allowed:
+        - pos. and neg. strategy cannot be set to 'semihard' simultaneously
+        - When pos. or neg. strategy is set to 'semihard' the other cannot be
+          set to 'all'
+
+        :param pos_strategy: Strategy for selecting positive samples
+        :param neg_strategy: Strategy for selecting negative samples
+        """
+        self.strategic_mining_helper = TorchStrategicMiningHelper(
+            pos_strategy, neg_strategy
+        )
+
+    def mine(
+        self, labels: paddle.Tensor, distances: paddle.Tensor
+    ) -> Tuple[paddle.Tensor, paddle.Tensor, paddle.Tensor]:
+        """Generate all possible triplets.
+
+        :param labels: A 1D tensor of item labels (classes)
+        :param distances: A tensor matrix of pairwise distance between each two item
+            embeddings
+
+        :return: three 1D tensors, holding the anchor index, positive index and
+            negative index of each triplet, respectively
+        """
+        assert len(distances) == len(labels)
+
+        # Needed, else sigfault if empty
+        if labels.size == 0:
+            return _empty_tensor(), _empty_tensor(), _empty_tensor()
+
+        labels1, labels2 = labels.unsqueeze(1), labels.unsqueeze(0)
+        matches = paddle.cast(labels1 == labels2, 'int32')
+        diffs = 1 - matches
+
+        # Apply mining strategy
+        updated_matches, updated_diffs = self.strategic_mining_helper.apply_strategy(
+            torch.Tensor(matches.numpy()),
```

Comment:
Using `Tensor` instead of `tensor` will convert everything to `float32` - but this is not the desired type, we want `uint8`. Also, you are not setting the device